Source/WebCore/Modules/push-api/PushEvent.h

```diff
@@ -37,14 +39,32 @@ class PushEvent final : public ExtendableEvent {
 public:
     static Ref<PushEvent> create(const AtomString&, PushEventInit&&, IsTrusted = IsTrusted::No);
     static Ref<PushEvent> create(const AtomString&, ExtendableEventInit&&, std::optional<Vector<uint8_t>>&&, IsTrusted);
+    static Ref<PushEvent> create(const AtomString&, ExtendableEventInit&&, Ref<Notification>, std::optional<uint64_t> appBadge, IsTrusted);
     ~PushEvent();
 
     PushMessageData* data() { return m_data.get(); }
+    Notification* notification();
+    std::optional<uint64_t> appBadge();
+
+    Notification* proposedNotification() const { return m_proposedNotification.get(); }
+    std::optional<uint64_t> proposedAppBadge() const { return m_proposedAppBadge; }
+
+    void setUpdatedNotification(Notification* notification) { m_updatedNotification = notification; }
+    std::optional<NotificationData> updatedNotificationData() const;
+
+    void setUpdatedAppBadge(std::optional<uint64_t>&& updatedAppBadge) { m_updatedAppBadge = WTFMove(updatedAppBadge); }
+    const std::optional<std::optional<uint64_t>>& updatedAppBadge() const { return m_updatedAppBadge; }
 
 private:
-    PushEvent(const AtomString&, ExtendableEventInit&&, std::optional<Vector<uint8_t>>&&, IsTrusted);
+    PushEvent(const AtomString&, ExtendableEventInit&&, std::optional<Vector<uint8_t>>&&, RefPtr<Notification>, std::optional<uint64_t> appBadge, IsTrusted);
 
     RefPtr<PushMessageData> m_data;
+
+    RefPtr<Notification> m_proposedNotification;
+    std::optional<uint64_t> m_proposedAppBadge;
+
+    RefPtr<Notification> m_updatedNotification;
+    std::optional<std::optional<uint64_t>> m_updatedAppBadge;
```

Comment:
I am curious why this needs to be of the type `std::optional<std::optional<uint64_t>>`?

Reply:
LOL - Second time to have talked about this, first when originally implemented:

An app badge is "an optional uint64_t" - You either set it to a number, or you set it to nothing to clear it.
And in this code, we need to capture "has the app badge been overridden at all?"

I could create a `struct AppBadge` which has a `uint64_t` member and a `bool m_isNull` member, and then we'd have `std::optional<AppBadge>` here instead, and the code would read nicer

Reply:
In fact if I have to update this patch again for EWS, I might add `typedef std::optional<uint64_t> AppBadge` for readability.